File: LeetCode/78.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        self.res = []
        def backtrack(i, cand=[]):
            if i == len(nums):
                self.res.append(cand)
                return
            backtrack(i+1, cand+[nums[i]])
            backtrack(i+1, cand)
        
        backtrack(0)
        return self.res

File: leetcode-new/78.cpp
class Solution
{
public:
    vector<vector<int>> subsets(vector<int> &nums)
    {
        vector<vector<int>> res;
        vector<int> cand;
        backtrack(nums, 0, res, cand);
        return res;
    }
    void backtrack(vector<int> &nums, int i, vector<vector<int>> &res, vector<int> &cand)
    {
        if (i == nums.size())
        {
            res.push_back(cand);
            return;
        }
        cand.push_back(nums[i]);
        backtrack(nums, i + 1, res, cand);
        cand.pop_back();
        backtrack(nums, i + 1, res, cand);
    }
};
